### PR TITLE
8324933: ConcurrentHashTable::statistics_calculate synchronization is expensive

### PR DESCRIPTION
**Pull request by fisk**
In the ConcurrentHashTable::statistics_calculate function, we enter and exit a ScopedCS with the global counter for every single bucket. This has showed up to be pretty intense on some machines. We should make the synchronization a bit less intense here. This patch adds simple batching so we synchronize once per 128 buckets instead of every single one.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324933](https://bugs.openjdk.org/browse/JDK-8324933): ConcurrentHashTable::statistics_calculate synchronization is expensive (**Enhancement** - P4)

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [3b26cf9c](https://git.openjdk.org/jdk/pull/17629/files/3b26cf9c95de14c3e4b90162e32801a823244f76)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [3b26cf9c](https://git.openjdk.org/jdk/pull/17629/files/3b26cf9c95de14c3e4b90162e32801a823244f76)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17629/head:pull/17629` \
`$ git checkout pull/17629`

Update a local copy of the PR: \
`$ git checkout pull/17629` \
`$ git pull https://git.openjdk.org/jdk.git pull/17629/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17629`

View PR using the GUI difftool: \
`$ git pr show -t 17629`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17629.diff">https://git.openjdk.org/jdk/pull/17629.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17629#issuecomment-1916583216)

### CODE REVIEW

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp [hunk 1]
```diff
@@ -1222,23 +1222,38 @@ template <typename VALUE_SIZE_FUNC>
...
+  size_t num_batches = table->_size / batch_size;
+  for (size_t current_batch = 0; current_batch <= num_batches; current_batch++) {
+    size_t batch_start = current_batch * batch_size;
+    size_t batch_end;
+    if (current_batch == num_batches) {
+      // Last batch; walk over the remaining part of the table
+      batch_end = batch_start + table->_size % batch_size;
+    } else {
+      // Not last batch; walk over the current batch
+      batch_end = batch_start + batch_size;
```
Comment (tschatzl) [inline src/hotspot/share/utilities/concurrentHashTable.inline.hpp:1238]: Something like (untested):
```suggestion
  for (size_t start_batch = 0; start_batch <= _table->_size; start_batch += batch_size {
    size_t batch_end = MIN2(start_batch + batch_size, _table->_size);
```

seems to be much much easier to follow (and corresponds to "usual" code for such loops) than the suggested one.
For extra performance `_table->_size` could be hoisted as well, but the compiler may alread do that anyway.

**Review by tschatzl — CHANGES_REQUESTED** [review]

**Comment by fisk** [comment]
Thanks for the review @tschatzl! I liked your suggested change. It looks simpler now.

**Review by coleenp — APPROVED** [review]
This seems reasonable, as long as the critical section is held throughout a bucket iteration.

**Review by tschatzl — APPROVED** [review]
lgtm. A comment why we need to do the batched iteration (without looking at versioning system logs) would be nice.

**Comment by fisk** [comment]
Thanks for the reviews, @coleenp and @tschatzl! I will add a comment before integrating.

**Pull request closed without merging**